h/services/group.py

```diff
@@ -119,6 +122,57 @@ def groupids_created_by(self, user):
             g.pubid for g in self.session.query(Group.pubid).filter_by(creator=user)
         ]
 
+    @classmethod
+    def _group_filter(cls, query, pubids_or_groupids: List[str]):
+        """
+        Apply a filter to a query to efficiently limit it to certain groups.
+
+        This works for a mixed list of pubids or groupids.
+        """
+
+        # We will build up filter clauses for a single efficient query
+        clauses = []
+
+        if pubids := [
+            id_ for id_ in pubids_or_groupids if not group_util.is_groupid(id_)
+        ]:
+            clauses.append(cls._in_or_equals(Group.pubid, pubids))
+
+        if groupids := [
+            id_ for id_ in pubids_or_groupids if group_util.is_groupid(id_)
+        ]:
+            by_authority = defaultdict(list)
+            for parts in (group_util.split_groupid(groupid) for groupid in groupids):
+                by_authority[parts["authority"]].append(parts["authority_provided_id"])
+
+            # We'll probably only ever get called with a single authority, but
+            # there's no harm in being able to support multiple
+            for authority, authority_provided_ids in by_authority.items():
+                clauses.append(
+                    sa.and_(
+                        Group.authority == authority,
+                        cls._in_or_equals(
+                            Group.authority_provided_id, authority_provided_ids
+                        ),
+                    )
+                )
+
+        if len(clauses) == 1:
+            query = query.filter(clauses[0])
+        elif clauses:
+            query = query.filter(sa.or_(*clauses))
+
+        return query
+
+    @staticmethod
+    def _in_or_equals(field, values):
+        """Get an IN clause for multiple values or an equals clause for one."""
+
+        if len(values) == 1:
+            return field == values[0]
+
+        return field.in_(values)
```

Comment:
A tiny further optimisation. I'm not sure if Postgres is smart enough to do this sort of thing for itself during query optimisation (quite possibly), but it's not a lot of code for us to be _sure_ we are emitting queries which are about as efficient as they can be.

With this in place, the query produced by this is equally effective at retrieving a single value as many.

Reply:
I think it does actually: https://stackoverflow.com/questions/44761351/potsgres-performance-where-versus-where-in-for-single-values

I'll simplify this down